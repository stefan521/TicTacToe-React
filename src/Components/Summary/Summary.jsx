/* eslint-disable class-methods-use-this */
import React from 'react';

class Summary extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      descending: true,
    };
  }

  getPlaySummary(move) {
    if (move.player === null) {
      return 'go to game start';
    }

    const summary = `go to move ${move.player} at `;
    const x = move.square % 3;
    let y = Math.floor(move.square / 3);
    // the origin of the xy axis is top left, not bottom left
    y = 2 - y;
    const coordinates = `(${x}, ${y})`;
    return (summary + coordinates);
  }

  getMoveAtIndex(index) {
    const { descending } = this.state;
    const { moves } = this.props;

    if (descending) {
      return index;
    }
    const reversedIndex = moves - index - 1;
    return reversedIndex;
  }

  getMoves() {
    const { descending } = this.state;
    const { moves } = this.props;
    const { onClick } = this.props;
    const { stepNumber } = this.props;

    const orderedMoves = descending ? moves : moves.slice().reverse();
    // make a li for each move
    return orderedMoves.map((move, index) => {
      const summary = this.getPlaySummary(move);
      return (
        // eslint-disable-next-line react/no-array-index-key
        <li key={index}>
          <button
            type="button"
            onClick={() => onClick(this.getMoveAtIndex(index))}
            style={(this.getMoveAtIndex(index) === stepNumber)
              ? { fontWeight: 'bold' } : { fontWeight: 'normal' }}
          >
            {summary}
          </button>
        </li>
      );
    });
  }

  reverseMoves() {
    const { descending } = this.state;

    this.setState({
      descending: !descending,
    });
  }

  render() {
    const { descending } = this.state;

    return (
      <div>
        <button
          type="button"
          onClick={() => this.reverseMoves()}
        >
          { `view moves ${descending ? 'descending' : 'ascending'}` }
        </button>
        <ol reversed={!descending}>
          {this.getMoves()}
        </ol>
      </div>
    );
  }
}

export default Summary;
