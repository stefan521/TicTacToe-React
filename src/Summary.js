import React from 'react';

export default class Summary extends React.Component {

  getPlaySummary(move, index) {
    if(index === 0) {
      return "go to game start";
    }

    let summary = "go to move " + move.player + " at ";
    let x = move.square % 3;
    let y = Math.floor(move.square / 3);
    //the origin of the xy axis is top left, not bottom left
    y = 2 - y;
    let coordinates = "(" + x + ", " + y + ")"
    return (summary + coordinates);
  }

  getMoves() {
    const moves = this.props.moves.map((move, index) => {
      const desc = this.getPlaySummary(move,index);
        return (
          <li key={index}>
            <button
              onClick={() => this.props.onClick(index)}
              style={(index === this.props.stepNumber) ? {fontWeight: "bold"} :
                     {fontWeight: "normal"}}
            >
              {desc}
            </button>
          </li>
        );
    });

    return moves;
  }

  render() {
    return(
      <div>
        <ol>{this.getMoves()}</ol>
      </div>
    );
  }
}
