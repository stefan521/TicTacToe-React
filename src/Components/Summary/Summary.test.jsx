import React from 'react';
import { shallow, mount } from 'enzyme';
import Summary from './Summary.jsx';
import { wrap } from 'module';

describe('Status', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<Summary 
      moves={
        [
          {
            player: 'X',
            square: 0,
          },
          {
            player:
            'O',
            square: 1,
          },
        ]
      }
      onClick={jest.fn()}
      stepNumber={2}
      xIsNext={true}
      status={
        {
          gameEnded: false,
          winner: null,
          line: null,
        }
      }
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('reverses move order', () => {
    const wrapper = mount(<Summary 
      moves={
        [
          {
            player: 'X',
            square: 0,
          },
          {
            player:
            'O',
            square: 1,
          },
        ]
      }
      onClick={jest.fn()}
      stepNumber={2}
      xIsNext={true}
      status={
        {
          gameEnded: false,
          winner: null,
          line: null,
        }
      }
    />);

    expect(wrapper.find('button').length).toBe(3);
    expect(wrapper.find('button').at(0).text()).toEqual('view moves descending');
    expect(wrapper.find('button').at(1).text()).toEqual('go to move X at (0, 2)');
    expect(wrapper.find('button').at(2).text()).toEqual('go to move O at (1, 2)');

    const buttonReverse = wrapper.find('#button-reverse');
    buttonReverse.simulate('click');

    expect(wrapper.find('button').length).toBe(3);
    expect(wrapper.find('button').at(0).text()).toEqual('view moves ascending');
    expect(wrapper.find('button').at(1).text()).toEqual('go to move O at (1, 2)');
    expect(wrapper.find('button').at(2).text()).toEqual('go to move X at (0, 2)');
  });
});
