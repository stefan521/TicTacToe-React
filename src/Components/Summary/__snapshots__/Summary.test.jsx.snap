// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status matches snapshot 1`] = `
<div
  className="game-info"
>
  <div
    className="next-player"
  >
    <Status
      status={
        Object {
          "gameEnded": false,
          "line": null,
          "winner": null,
        }
      }
      xIsNext={true}
    />
  </div>
  <div
    className="history"
  >
    <button
      id="button-reverse"
      onClick={[Function]}
      type="button"
    >
      view moves descending
    </button>
    <button
      className="btn-history"
      key="0"
      onClick={[Function]}
      type="button"
    >
      go to move X at (0, 2)
    </button>
    <button
      className="btn-history"
      key="1"
      onClick={[Function]}
      type="button"
    >
      go to move O at (1, 2)
    </button>
  </div>
</div>
`;
